### ISSUE TITLE
ESC does not open the split console if focus is in debugger editor

### ISSUE THREAD

**Issue by jnachtigall** [devtools-html/debugger.html#1404]
Similar to #1367, but since it's not a debugger but global devtool's shortcut (and also less important) I'm adding as a new ticket: 

The ESC key (which toggles the split console) also does not work in the new debugger. Works in all other tabs like network or inspector, only in the new debugger pressing ESC does nothing. Windows 10, Firefox 52.

@clarkbw Mind looking into this as well?

**Comment by clarkbw** [comment]
Yeah, this works on the Mac so perhaps something with Windows broke.  

I'm traveling right now and won't be back at my Windows box until later next week.  So if someone else could look into it sooner that would be nice.

**Comment by jasonLaster** [comment]
Hmm, we have this test which should catch this:

https://github.com/devtools-html/debugger.html/blob/master/src/test/mochitest/browser_dbg-console.js#L29-L33

I believe the toolbox handles showing the console splitbox, so perhaps @juliandescottes or @gregtatum would have some insights.

**Comment by jnachtigall** [comment]
Just tested: Works in Ubuntu, too.

**Comment by jasonLaster** [comment]
Thanks. Strange that we've got these windows events issues

**Comment by juliandescottes** [comment]
@jnachtigall  on Windows 10 with FF52 dev-edition I have the following behavior:
* if the focus is in the debugger editor, ESC does not open the split console
* if the focus is anywhere else in the debugger, ESC opens the split console

However I have the same behavior on OSX. Is ESC doing nothing even if you have the focus outside of the editor?

**Comment by jnachtigall** [comment]
@juliandescottes It's exactly as you describe. On Ubuntu and Windows. Thanks for nailing it done. 

The first time I noticed this, I was debugging at a breakpoint on Windows (at work) which obviously focuses in the debugger editor. Later on when quickly retesting with Ubuntu (at home) I did not care to set a breakpoint or look into where I set the focus.

**Comment by jasonLaster** [comment]
So, here are the two handlers I know off the top of my head:

* the search bar seems to not stop bubbling
* autocomplete was *fixed* earlier :)
* the best way to test this is to add an event handler to **window** and see if you can get this to be called... 

```diff
diff --git a/src/components/EditorSearchBar.js b/src/components/EditorSearchBar.js
index 75c3a53..d1e91b0 100644
--- a/src/components/EditorSearchBar.js
+++ b/src/components/EditorSearchBar.js
@@ -85,6 +85,7 @@ const EditorSearchBar = React.createClass({
   },

   onEscape(e) {
+    // => one handler
     this.closeSearch(e);
   },

diff --git a/src/components/SourceSearch.js b/src/components/SourceSearch.js
index a5d9fa6..cbdf96e 100644
--- a/src/components/SourceSearch.js
+++ b/src/components/SourceSearch.js
@@ -73,6 +73,7 @@ const Search = React.createClass({
   },

   onEscape(shortcut, e) {
+    // => another handler
     if (this.props.searchOn) {
       e.preventDefault();
       this.setState({ inputValue: "" });
```

**Comment by juliandescottes** [comment]
Investigated a bit, the issue comes from the fact that the escape key is used as a Shortcut for the "singleSelection" command. As soon as a key has any handler defined, code mirror will e.preventDefault() on the keydown event, which prevents the keypress event from being triggered, which is used by the toolbox to toggle the split console.

This is only useful when multiple selections are used, so I propose we disable the default keymap and add a key shortcut on Esc that does the same, except it only calls prevent default if the selection was modified by the handler.

**Comment by jasonLaster** [comment]
That sounds good.

We could also re-trigger the event. I wonder why this works for the old
debugger? Sane editor....

**Comment by juliandescottes** [comment]
@ I wonder why this works for the old debugger?

The old debugger is already setting the Escape key shortcut to false in http://searchfox.org/mozilla-central/source/devtools/client/debugger/debugger-view.js#250 

I don't think they go the extra mile and try to restore the behaviour as I suggested though.

**Comment by jasonLaster** [comment]
This was fixed in #1415
